src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
```

Comment:
As the size field is read and written to by every `push` and `pop` it becomes a sequential bottleneck.

I wonder whether it would be possible to adapt the technique I described [here](https://github.com/roburio/miou/issues/2#issuecomment-1630641505) for a lock-free MS queue to get less contentious length counting for the lock-based version.